client/app/cloud/project/compute/regions/regions.controller.js

```diff
@@ -0,0 +1,90 @@
+class RegionsCtrl {
+  constructor(
+    $stateParams,
+    CloudMessage,
+    ServiceHelper,
+    ControllerHelper,
+    OvhApiCloudProjectRegion,
+    CloudProjectVirtualMachineAddService,
+    RegionService,
+  ) {
+    this.serviceName = $stateParams.projectId;
+    this.CloudMessage = CloudMessage;
+    this.ServiceHelper = ServiceHelper;
+    this.ControllerHelper = ControllerHelper;
+    this.OvhApiCloudProjectRegion = OvhApiCloudProjectRegion;
+    this.VirtualMachineAddService = CloudProjectVirtualMachineAddService;
+    this.RegionService = RegionService;
+  }
+
+  $onInit() {
+    this.availableRegionToAdd = null;
+    this.initLoaders();
+  }
+
+  initLoaders() {
+    this.regions = this.ControllerHelper.request.getHashLoader({
+      loaderFunction: () => this.OvhApiCloudProjectRegion.v6()
+        .query({ serviceName: this.serviceName })
+        .$promise
+        .then(regionIds => _.map(regionIds, region => this.RegionService.getRegion(region)))
+        .catch(error => this.ServiceHelper.errorHandler('cpci_add_regions_get_regions_error')(error)),
+    });
+    this.regions.load();
+    this.regionsByDatacenter = this.ControllerHelper.request.getHashLoader({
+      loaderFunction: () => this.regions
+        .promise
+        .then(regions => this.VirtualMachineAddService.constructor.groupRegionsByDatacenter(
+          regions,
+        )),
+    });
+    this.regionsByDatacenter.load();
+    this.regionsByContinent = this.ControllerHelper.request.getHashLoader({
+      loaderFunction: () => this.regionsByDatacenter
+        .promise
+        .then(regions => _.groupBy(regions, 'continent')),
+    });
+    this.regionsByContinent.load();
+
+    // load available regions
+    this.availableRegions = this.ControllerHelper.request.getHashLoader({
+      loaderFunction: () => this.OvhApiCloudProjectRegion.AvailableRegions().v6()
+        .query({ serviceName: this.serviceName })
+        .$promise
+        .then(regionIds => _.map(regionIds, region => this.RegionService.getRegion(region.name)))
+        .catch(error => this.ServiceHelper.errorHandler('cpci_add_regions_get_available_regions_error')(error)),
+    });
+    this.availableRegions.load();
+    this.availableRegionsByDatacenter = this.ControllerHelper.request.getHashLoader({
+      loaderFunction: () => this.availableRegions
+        .promise
+        .then(regions => this.VirtualMachineAddService.constructor.groupRegionsByDatacenter(
+          regions,
+        )),
+    });
+    this.availableRegionsByDatacenter.load();
+    this.availableRegionsByContinent = this.ControllerHelper.request.getHashLoader({
+      loaderFunction: () => this.availableRegionsByDatacenter
+        .promise
+        .then(regions => _.groupBy(regions, 'continent')),
+    });
+    this.availableRegionsByContinent.load();
+  }
```

Comment:
Maybe all those calls could be splited into smaller functions and maybe chained if they're all promises ?

Reply:
I have made as small as possible.

Reply:
I have split this to 6 functions